pkg/conversion/mapi2capi/aws.go

```diff
@@ -167,10 +167,17 @@ func (m *awsMachineSetAndInfra) ToMachineSetAndMachineTemplate() (*capiv1.Machin
 	capaMachineTemplate := awsMachineToAWSMachineTemplate(capaMachine, m.machineSet.Name, capiNamespace)
 
 	capiMachineSet, machineSetErrs := fromMAPIMachineSetToCAPIMachineSet(m.machineSet)
-	errs = append(errs, machineSetErrs.Errors()...)
+	if machineSetErrs != nil {
+		errs = append(errs, machineSetErrs.Errors()...)
+	}
 
 	capiMachineSet.Spec.Template.Spec = capiMachine.Spec
 
+	// We have to merge these two maps so that labels and annotations added to the template objectmeta are persisted
+	// along with the labels and annotations from the machine objectmeta.
+	capiMachineSet.Spec.Template.ObjectMeta.Labels = mergeMaps(capiMachineSet.Spec.Template.ObjectMeta.Labels, capiMachine.Labels)
```

Comment:
As `mergeMaps` is written currently, if a label exists in both the template and the machine, the machine's value will override the template's.

If anything else is acting on the machine, this could cause drift. Should we be concerned about that, or are the corresponding VAP (and probably MAPs later) going to be enough?

Reply:
If you were to look at L75 of this file, you can see where we are creating the Machine that gets converted here. We don't populate the Machine labels, so, there should be nothing in the list unless it is something we have moved from the Machine.Spec.ObjectMeta up to the Machine.ObjectMeta. That is, the labels that are managed by CAPI.

So, I think it's correct that these take priority, that said, it would find it weird if the MAPI MachineSet had a label at the Machine level that was then different at the node level right?

Reply:
That would be weird. As long as this is intentional 👍